File: scripts/reprocess_car_image.py
#!/usr/bin/env python3
"""
차량 이미지 재처리 스크립트
소스: Supabase URL(360 프레임) 또는 로컬 백업 파일
출력: 800×600 흰 배경, 차량 크기/위치 정규화

사용법:
  python3 scripts/reprocess_car_image.py \
    --source <URL or path> \
    --output <output.webp> \
    --width-ratio 72 \
    --v-position 55
"""

import argparse
import io
import sys
import urllib.request
from typing import Optional, Tuple
import numpy as np
from PIL import Image

try:
    from rembg import remove as rembg_remove
    REMBG_AVAILABLE = True
except ImportError:
    REMBG_AVAILABLE = False


CANVAS_W, CANVAS_H = 800, 600


def load_image(source: str) -> Image.Image:
    if source.startswith('http://') or source.startswith('https://'):
        with urllib.request.urlopen(source, timeout=30) as resp:
            data = resp.read()
        return Image.open(io.BytesIO(data)).convert('RGBA')
    else:
        return Image.open(source).convert('RGBA')


def remove_background(img: Image.Image) -> Image.Image:
    if not REMBG_AVAILABLE:
        print('[warn] rembg 없음 — 배경 제거 건너뜀', file=sys.stderr)
        return img
    buf = io.BytesIO()
    img.save(buf, format='PNG')
    buf.seek(0)
    result = rembg_remove(buf.read())
    return Image.open(io.BytesIO(result)).convert('RGBA')


def get_bbox(img_rgba: Image.Image, threshold: int = 235) -> Optional[Tuple[int, int, int, int]]:
    arr = np.array(img_rgba)
    alpha = arr[:, :, 3]
    rgb = arr[:, :, :3]
    is_opaque = alpha > 10
    not_white = np.any(rgb < threshold, axis=2)
    mask = is_opaque & not_white
    rows = np.any(mask, axis=1)
    cols = np.any(mask, axis=0)
    if not rows.any() or not cols.any():
        return None
    row_idx = np.where(rows)[0]
    col_idx = np.where(cols)[0]
    if len(row_idx) == 0 or len(col_idx) == 0:
        return None
    rmin = row_idx[0].item()
    rmax = row_idx[-1].item()
    cmin = col_idx[0].item()
    cmax = col_idx[-1].item()
    return cmin, rmin, cmax, rmax


def place_on_canvas(car: Image.Image, bbox: Tuple[int, int, int, int], width_ratio: float, v_position: float) -> Image.Image:
    cx1, cy1, cx2, cy2 = bbox
    car_w = cx2 - cx1
    car_h = cy2 - cy1

    target_w = int(CANVAS_W * width_ratio / 100)
    max_h = int(CANVAS_H * 0.75)

    scale = min(target_w / max(car_w, 1), max_h / max(car_h, 1))
    new_w = int(car_w * scale)
    new_h = int(car_h * scale)

    cropped = car.crop((cx1, cy1, cx2, cy2))
    resized = cropped.resize((new_w, new_h), Image.LANCZOS)

    canvas = Image.new('RGB', (CANVAS_W, CANVAS_H), (255, 255, 255))
    paste_x = (CANVAS_W - new_w) // 2
    center_y = int(CANVAS_H * v_position / 100)
    paste_y = center_y - new_h // 2
    paste_y = max(0, min(paste_y, CANVAS_H - new_h))

    canvas.paste(resized, (paste_x, paste_y), resized)
    return canvas


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--source', required=True)
    parser.add_argument('--output', required=True)
    parser.add_argument('--width-ratio', type=float, default=72)
    parser.add_argument('--v-position', type=float, default=55)
    args = parser.parse_args()

    print(f'소스 로드: {args.source}')
    img = load_image(args.source)

    print('배경 제거 중...')
    img_nobg = remove_background(img)

    bbox = get_bbox(img_nobg)
    if bbox is None:
        print('[warn] 차량 감지 실패 — 원본 전체 기준으로 배치합니다', file=sys.stderr)
        bbox = (0, 0, img_nobg.width, img_nobg.height)

    print(f'캔버스 합성 (가로 {args.width_ratio}%, 세로위치 {args.v_position}%)')
    canvas = place_on_canvas(img_nobg, bbox, args.width_ratio, args.v_position)

    canvas.save(args.output, 'WEBP', quality=90)
    print(f'저장 완료: {args.output}')


if __name__ == '__main__':
    main()

File: src/lib/reprocess-car-image.ts
import sharp from 'sharp';
import { readFile } from 'fs/promises';

const CANVAS_W = 800;
const CANVAS_H = 600;

async function loadImageBuffer(source: string): Promise<Buffer> {
  if (source.startsWith('http://') || source.startsWith('https://')) {
    const res = await fetch(source, { signal: AbortSignal.timeout(30_000) });
    if (!res.ok) throw new Error(`이미지 다운로드 실패: ${res.status} ${source}`);
    return Buffer.from(await res.arrayBuffer());
  }
  return readFile(source);
}

// Python get_bbox()와 동일 로직: 불투명 + 비흰색 픽셀 기준 bounding box
function getBbox(
  data: Uint8Array,
  width: number,
  height: number,
  threshold = 235
): [number, number, number, number] | null {
  let rmin = height, rmax = -1, cmin = width, cmax = -1;

  for (let row = 0; row < height; row++) {
    for (let col = 0; col < width; col++) {
      const i = (row * width + col) * 4;
      const alpha = data[i + 3];
      const r = data[i], g = data[i + 1], b = data[i + 2];
      // is_opaque & not_white
      if (alpha > 10 && (r < threshold || g < threshold || b < threshold)) {
        if (row < rmin) rmin = row;
        if (row > rmax) rmax = row;
        if (col < cmin) cmin = col;
        if (col > cmax) cmax = col;
      }
    }
  }

  if (rmax === -1) return null;
  return [cmin, rmin, cmax, rmax];
}

export async function reprocessCarImage(opts: {
  source: string;
  widthRatio: number;  // 50~90
  vPosition: number;   // 40~70
}): Promise<Buffer> {
  const { source, widthRatio, vPosition } = opts;

  console.log(`소스 로드: ${source}`);
  const inputBuf = await loadImageBuffer(source);

  // RGBA raw pixel 추출
  const { data, info } = await sharp(inputBuf)
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const { width, height } = info;

  // bounding box 감지
  let bbox = getBbox(new Uint8Array(data), width, height);
  if (!bbox) {
    console.warn('[warn] 차량 감지 실패 — 원본 전체 기준으로 배치합니다');
    bbox = [0, 0, width, height];
  }

  const [cx1, cy1, cx2, cy2] = bbox;
  const carW = cx2 - cx1;
  const carH = cy2 - cy1;

  // Python place_on_canvas()와 동일한 스케일 계산
  const targetW = Math.floor(CANVAS_W * widthRatio / 100);
  const maxH = Math.floor(CANVAS_H * 0.75);
  const scale = Math.min(targetW / Math.max(carW, 1), maxH / Math.max(carH, 1));
  const newW = Math.round(carW * scale);
  const newH = Math.round(carH * scale);

  const pasteX = Math.floor((CANVAS_W - newW) / 2);
  const centerY = Math.floor(CANVAS_H * vPosition / 100);
  const pasteY = Math.min(Math.max(centerY - Math.floor(newH / 2), 0), CANVAS_H - newH);

  console.log(`캔버스 합성 (가로 ${widthRatio}%, 세로위치 ${vPosition}%)`);

  // 차량 영역 crop → resize → 흰 배경 800×600에 합성
  const carBuf = await sharp(inputBuf)
    .ensureAlpha()
    .extract({ left: cx1, top: cy1, width: carW, height: carH })
    .resize(newW, newH, { fit: 'fill', kernel: 'lanczos3' })
    .toBuffer();

  const resultBuf = await sharp({
    create: {
      width: CANVAS_W,
      height: CANVAS_H,
      channels: 3,
      background: { r: 255, g: 255, b: 255 },
    },
  })
    .composite([{ input: carBuf, left: pasteX, top: pasteY }])
    .webp({ quality: 90 })
    .toBuffer();

  return resultBuf;
}
